src/Interpreters/MonotonicityCheckVisitor.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <DataTypes/DataTypeFactory.h>
+#include <Functions/FunctionFactory.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Interpreters/IdentifierSemantic.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Parsers/IAST.h>
+#include <Common/typeid_cast.h>
+
+namespace DB
+{
+
+using Monotonicity = IFunctionBase::Monotonicity;
+
+/// Checks from bottom to top if function composition is monotonous
+class MonotonicityCheckMatcher
+{
+public:
+    struct Data
+    {
+        const TablesWithColumns & tables;
+        const Context & context;
+        const std::unordered_map<String, ASTPtr> & group_by_function_hashes;
+        Monotonicity monotonicity{true, true, true};
+        ASTIdentifier * identifier = nullptr;
+        DataTypePtr arg_data_type = {};
+
+        void reject() { monotonicity.is_monotonic = false; }
+        bool isRejected() const { return !monotonicity.is_monotonic; }
+
+        bool canOptimize(const ASTFunction & ast_function) const
+        {
+            /// if GROUP BY contains the same function ORDER BY shouldn't be optimized
+            auto hash = ast_function.getTreeHash();
+            String key = toString(hash.first) + '_' + toString(hash.second);
+            if (group_by_function_hashes.count(key))
+                return false;
+
+            /// if ORDER BY contains aggregate function it shouldn't be optimized
+            if (AggregateFunctionFactory::instance().isAggregateFunctionName(ast_function.name))
+                return false;
+
+            return true;
+        }
+
+        bool getIdentifierAndType(const ASTFunction & ast_function)
+        {
+            identifier = ast_function.arguments->children[0]->as<ASTIdentifier>();
+            if (!identifier)
+                return false;
+
+            auto pos = IdentifierSemantic::getMembership(*identifier);
+            if (!pos)
+                pos = IdentifierSemantic::chooseTableColumnMatch(*identifier, tables, true);
+            if (!pos)
+                return false;
+
+            if (auto data_type_and_name = tables[*pos].columns.tryGetByName(identifier->shortName()))
+            {
+                arg_data_type = data_type_and_name->type;
+                return true;
+            }
+
+            return false;
+        }
+    };
+
+    static void visit(const ASTPtr & ast, Data & data)
+    {
+        if (const auto * ast_function = ast->as<ASTFunction>())
+            visit(*ast_function, data);
+    }
+
+    static void visit(const ASTFunction & ast_function, Data & data)
+    {
+        if (data.isRejected())
+            return;
+
+        auto arguments = ast_function.arguments;
+        if (arguments->children.size() != 1)
```

Comment:
Maybe add a comment that it doesn't work for functions with more than one argument.